ibis/expr/rules.py

```diff
@@ -237,7 +237,10 @@ def value(dtype, arg, **kwargs):
         # coerce python literal to ibis literal
         arg = ir.literal(arg)
 
-    if not isinstance(arg, ir.AnyValue):
+    if not isinstance(arg, ir.AnyValue) or type(arg) in (
```

Comment:
Could you please restore the previous check so we can keep the previous behavior? Eventually we should use abstract base classes to prevent creating untyped value expressions.

Reply:
I think there's value in preventing this kind of thing now, versus waiting until a refactor that may never happen. The current behavior of allowing `AnyValue`s is incorrect, until we have a valid use case for it.

Reply:
@kszucs Do you have any code that might rely on this behavior?  From what I understand, this is now basically incorrect and without this check the error message is just more incomprehensible.

Reply:
Fair point on the error message. 

My problem is that the actual type is determined by the underlying `datatype` and the `ir` types are merely shims to provide the python API. The value rules actually don't care about `ir` type, they check the `expr.type()` property which is a datatype, so `ir.AnyScalar(dt.int64)` would be valid according to the rules.

If we're not paranoid about the backward compatibility then it's fine to introduce this, though it requires a test case then.

Reply:
Thanks @kszucs, I think this next release will include several backward-incompatible changes so we don't need to be paranoid.  I'll add a test case to make sure the code is exercised.